hs/src/Reach/Optimize.hs

```diff
@@ -629,11 +675,12 @@ optLet at x e = do
                   _ ->
                     return ()
                 meh
+  svs <- asks eSvs
   case (extract x, (isPure e && canDupe e), e) of
-    (Just dv, True, _) -> doit dv
+    (Just dv, True, _) | not (S.member dv svs) -> doit dv
     -- Optimize arithmetic even if it is impure (PV_Safe).
     -- It may trap as an effect, but we don't need to worry about it happening multiple times
-    (Just dv, _, DLE_PrimOp {}) -> doit dv
+    (Just dv, _, DLE_PrimOp {}) | not (S.member dv svs) -> doit dv
```

Comment:
Name `not (S.member dv svs)` a function like `allowedToRemove` to make it intuitive